### DISCUSSION TITLE
Homomorphic Encryption of MariaDB contents in Appwrite Cloud

### DISCUSSION THREAD

**Discussion by zdanl** [appwrite/appwrite#7229] [Ideas]
The ability to scale is paramount and Appwrite Cloud for sure looks like a functional and profitable, viable business model. Unfortunately it's a bit like giving your users data to Firebase, neutralizing the ideals and promise of self-hosted and open source in general. 

I haven't thought this through, but is there a way to generate a cryptographic key upon first registration of the Appwrite Project Owner,  that he or she retains in the browser (Passkey API) or a Password Manager like Proton Pass; limiting Appwrite Employees ability to grasp the contents of a MariaDB even if they wanted to?

I realize most use cases, even querying, or displaying user owned data back in the app, prevent this categorically, but there is an academic concept called Homomorphic Encryption

> Homomorphic encryption is a form of
> encryption that allows computations to
 > be performed on encrypted data without
> first having to decrypt it. The resulting
> computations are left in an encrypted
> form which, when decrypted, result in an
> output that is identical to that produced
> had the operations been performed on
> the unencrypted data.

https://stackoverflow.com/questions/50467084/execution-of-homomorphic-encrypted-queries-on-databases

https://en.m.wikipedia.org/wiki/Homomorphic_encryption

Food for thought. 

![homo](https://github.com/appwrite/appwrite/assets/114028070/7144ef3b-ddaa-4f76-837e-a584fa8e5069)

**Comment by stnguyen90** [comment]
@zdanl, thanks for raising this!

> I haven't thought this through, but is there a way to generate a cryptographic key upon first registration of the Appwrite Project Owner, that he or she retains in the browser (Passkey API) or a Password Manager like Proton Pass; limiting Appwrite Employees ability to grasp the contents of a MariaDB even if they wanted to?

So, if the project owner only has the key, how would Appwrite decrypt any of the project data for any of the operations needed for end users. For example, If you set up OAuth2 integration with Google, you have to give Appwrite a client ID and secret. This client ID and secret needs to be used when end users log in for the OAuth2 flow.

**Reply by zdanl** [reply]
Yes, MariaDB at large was too far fetched, as it destroys the functionality that makes Appwrite so great. But single collections, possibly including the field types, possibly just the values, would be an interesting experiment, even more compelling if we build it with Web3Auth or SIWE (Ethereum Auth) via Browser Extension or web3 Browser. 

web3 is all about the users retaining ownership over their data.
